src/host/readDataCooked.cpp

```diff
@@ -173,24 +88,89 @@ COOKED_READ_DATA::COOKED_READ_DATA(_In_ InputBuffer* const pInputBuffer,
         //
         // In other words, it assumed that the number of code units in the initial data corresponds 1:1 to
         // the column count. This meant that the API never supported tabs for instance (nor wide glyphs).
-        // The new implementation still doesn't support tabs, but it does fix support for wide glyphs.
-        // That seemed like a good trade-off.
-
-        // NOTE: You can't just "measure" the length of the string in columns either, because previously written
-        // wide glyphs might have resulted in padding whitespace in the text buffer (see ROW::WasDoubleBytePadded).
-        // The alternative to the loop below is counting the number of padding glyphs while iterating backwards. Either approach is fine.
-        til::CoordType distance = 0;
-        for (size_t i = 0; i < initialData.size(); i = TextBuffer::GraphemeNext(initialData, i))
+        //
+        //
+        // The new implementation is a lot more complex to be a little more correct.
+        // It replicates part of the _redisplay() logic to layout the text at various
+        // starting positions until it finds one that matches the current cursor position.
+
+        const auto size = _screenInfo.GetVtPageArea().Dimensions();
+
+        // Guess the initial cursor position based on the string length, assuming that 1 char = 1 column.
+        const auto columnRemainder = gsl::narrow_cast<til::CoordType>((initialData.size() % size.width));
+        const auto bestGuessColumn = (cursorPos.x - columnRemainder + size.width) % size.width;
+
+        std::wstring line;
+        LayoutResult res;
+        til::CoordType bestDistance = til::CoordTypeMax;
+        til::CoordType bestColumnBegin = 0;
+        til::CoordType bestNewlineCount = 0;
+
+        line.reserve(size.width);
+
+        // Do a brute force search for the best starting position that ends at the current cursor position.
+        // The search is centered around `bestGuessColumn` with offsets 0, 1, -1, 2, -2, 3, -3, ...
+        for (til::CoordType i = 0, attempts = 2 * size.width; i <= attempts; i++)

```

Comment:
Oh right, I forgot to mention in the comment _why_ we need to do that! I'll update it.

We need to do this, because of wide glyphs. When a wide glyph intersects the end of a row, we insert a padding space. We're given an "end position" and a string and we need to find its starting position. The problem is that multiple starting positions can result in the same end position depending on what wide glyphs intersect the end of their row where. Because of that it's not trivially possible to layout those lines backwards starting from the end position.